### PR TITLE
Upgrade golang to 1.16

### PR DESCRIPTION
**Pull request by algojack**
To keep inline with security updates and go-algorand we are upgrading to golang 1.16

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -16,7 +16,7 @@ workflows:
           name: test_with_go_<< matrix.go_version >>
           matrix: &go-version-matrix
             parameters:
-              go_version: ["1.14.7", "1.15.15"]
+              go_version: ["1.16.11"]
```
Comment (cce) [inline .circleci/config.yml:19]: if you're replacing the two we used to test against with one, might as well test the two supported releases (Go 1.16.11 and go 1.17.4)

File: go.mod [hunk 2]
```diff
@@ -22,5 +22,4 @@ require (
 	github.com/spf13/pflag v1.0.5
 	github.com/spf13/viper v1.7.1
 	github.com/stretchr/testify v1.7.0
-	github.com/vektra/mockery v1.1.2 // indirect
```
Comment (algojack) [inline go.mod:25]: not sure about this one. Can someone confirm we no longer need it?

**Comment by cce** [comment]
One potential issue, this repo imports go-algorand as a submodule and uses it for the build, I think @tolikzinovyev would have an opinion on how to update the submodule to match the Go upgrade here

**Comment by algojack** [comment]
> One potential issue, this repo imports go-algorand as a submodule and uses it for the build, I think @tolikzinovyev would have an opinion on how to update the submodule to match the Go upgrade here

updated it, thanks

File: .circleci/config.yml (hunk 1 above)
Comment (algojack) [reply .circleci/config.yml]: added, thanks

**Comment by algojack** [comment]
seems to be failing on `go vet ./...` but when i run it locally, it passes just fine, and those go get commands don't do anything on local.

**Comment by cce** [comment]
I have a workaround for the go 1.16 build issue in e799997ddb306786e1d5db522aecfe74a980adca. Something like this might work too:

```diff
diff --git a/.circleci/config.yml b/.circleci/config.yml
index bcd3370..12a0f8f 100644
--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -28,6 +28,11 @@ jobs:
     steps:
       - go/install:
           version: << parameters.go_version >>
+      - when:
+          condition:
+            equal: [ "1.16.11", << parameters.go_version >> ]
+          steps:
+            - run: go env -w GOFLAGS=-mod=mod
       - install_dependencies
       - run_tests
       - upload_coverage
```

File: cmd/block-generator/docker/Dockerfile-generator [hunk 3]
```diff
@@ -1,4 +1,4 @@
-FROM golang:1.13.12-alpine
+FROM golang:1.16.11-alpine
```
Comment (tolikzinovyev) [inline cmd/block-generator/docker/Dockerfile-generator:1]: perhaps in docker we can use the newest version of go?

**Review by tolikzinovyev — APPROVED** [review]
Looks great!

File: cmd/block-generator/docker/Dockerfile-generator (hunk 3 above)
Comment (algojack) [reply cmd/block-generator/docker/Dockerfile-generator:1]: i'm all for moving to 1.17. We need to get managers approval on this. Can you guys talk to Tsachi and see if he approves going straight to 1.17 everywhere?

**Comment by algojack** [comment]
> I have a workaround for the go 1.16 build issue in [e799997](https://github.com/algorand/indexer/commit/e799997ddb306786e1d5db522aecfe74a980adca). Something like this might work too:...

Merged your commit and it worked. Thanks!

File: cmd/block-generator/docker/Dockerfile-generator (hunk 3 above)
Comment (tolikzinovyev) [reply cmd/block-generator/docker/Dockerfile-generator:1]: > i'm all for moving to 1.17. We need to get managers approval on this. Can you guys talk to Tsachi and see if he approves going straight to 1.17 everywhere?

@tsachiherman

**Review by winder — COMMENTED** [review]
Did you change the go-algorand submodule on purpose?

**Review by winder — CHANGES_REQUESTED** [review]
@algojack did you change the go-algorand submodule on purpose? Aside from that changes LGTM and I'm ready to merge the PR.

**Comment by algojack** [comment]
> @algojack did you change the go-algorand submodule on purpose? Aside from that changes LGTM and I'm ready to merge the PR.

Correct, on purpose, so sync with go-algorand go 1.16 update

**Review by winder — APPROVED** [review]

**Pull request merged**